Strftime in views is probably a mistake (#49)
andycroll/andycroll.com: _posts/2019/2019-06-23-using-strftime-in-a-rails-view-is-probably-a-mistake.md

---
title: "Using strftime in a Rails view is probably a mistake"
description: "I heard you like the % character"
layout: article
category: ruby
image:
  base: '2019/using-strftime-in-a-rails-view-is-probably-a-mistake'
  alt: 'Old broken clocks'
  credit:  Heather Zabriskie
  source: "https://unsplash.com/photos/yBzrPGLjMQw"

---

The `strftime` method appears in many languages, [all the way back to C](https://en.wikipedia.org/wiki/C_date_and_time_functions#strftime)!

> The syntax of the (mostly impenetrable) time formatting arguments haven’t changed that much in years and are comprehensively covered in the [Ruby documentation for strftime](https://ruby-doc.org/core-2.6.3/Time.html#method-i-strftime). They're remarkably consistent across languages!

Reviewer: I wonder if it's worth keeping stuff to do with other languages all together: ie it appears in many languages _and_ it's remarkably consistent AND the arguments haven't changed much and you can find them here.

I'd suggest starting by saying it appears in many places and it's consistent. And then introducing the Ruby docs.

However when `strftime` is used in your views, there's a high chance it will lead to confusion and inconsistency.

## Instead of…

…using `strftime` in your views to format dates and times:

```erb
<%= @user.last_signed_in_at.strftime("%m-%e-%y %H:%M") %>
```

## Use…

…the built in Rails time and date formats. Or add to them, like I do here, creating my own `:stamp` format for `Date` and `Time`.

### `config/initializers/time_formats.rb`

```ruby
Date::DATE_FORMATS[:stamp] = "%Y%m%d" # YYYYMMDD
Time::DATE_FORMATS[:stamp] = "%Y%m%d%H%M%S" # YYYYMMDDHHMMSS
```

> …then in the views.

Reviewer: *Then, in the views:

```erb
<%= @user.last_signed_in_at.to_s(:stamp) %>
```

## Why?

By using a method with confusing and complex arguments in your views, you’re setting yourself up for inconsistency.

> The main benefit for setting up global application date and time formats is for the developers who come after you, including _future_ you. The result is a small number of consistent and pre-determined ways to present time throughout your application, and future developers can be spared building their own formats!

Reviewer: *defining/establishing — you used 'setting...up' in the paragraph before.

Users benefit from this consistency too. If dates and times are always presented in the same way, it takes less time for them to parse. I.e. Ensuring you always say “Jun 11” rather than “11 Jun”; It's a small thing, but worth sweating the details.

> There’s a couple of useful sites that can help you get exactly the date formatting you prefer:

Reviewer: ```suggestion
Here's a couple of useful sites that can help you get to exactly the date formatting you prefer:
```

[For A Good Strftime](https://www.foragoodstrftime.com) contains a visually pleasant reference for `strftime` formatting strings and lets you build a live date format you can use application-wide.

> [Strftimer](http://strftimer.com) lets you paste in a real string representation of a date of time and returns the `strftime` date format string that would produce that formatting.

Reviewer: *date or time

> I can never remember the `Date` and `Time` formats that are provided by default, plus they were hard to find in the documents, so _I_ built a site that spells out the [Rails Date and Time formats](https://railsdatetimeformats.com).

Reviewer: OH!! That was not clear last time. 😂

## Why not?

> There’s no real performance harm in calling `strftime` in views. This is about about organisation and consistency.

Reviewer: Double 'about'!

---
title: "Using strftime in a Rails view is probably a mistake"
description: "I heard you like the % character"
layout: article
category: ruby
image:
  base: '2019/using-strftime-in-a-rails-view-is-probably-a-mistake'
  alt: 'Old broken clocks'
  credit:  Heather Zabriskie
  source: "https://unsplash.com/photos/yBzrPGLjMQw"

---

The `strftime` method appears in many languages, [all the way back to C](https://en.wikipedia.org/wiki/C_date_and_time_functions#strftime)!

The syntax of the (mostly impenetrable) time formatting arguments haven’t changed that much in years and are comprehensively covered in the [Ruby documentation for strftime](https://ruby-doc.org/core-2.6.3/Time.html#method-i-strftime). They're remarkably consistent across languages!

> However when `strftime` is used in your views it’s a possible location for confusion and inconsistency.

Reviewer: ```suggestion
However when `strftime` is used in your views, there's a high chance it will lead to confusion and inconsistency.
```

## Instead of…

…using `strftime` in your views to format dates and times:

```erb
<%= @user.last_signed_in_at.strftime("%m-%e-%y %H:%M") %>
```

## Use…

…the built in Rails time and date formats. Or add to them.

### `config/initializers/time_formats.rb`

```ruby
Date::DATE_FORMATS[:stamp] = "%Y%m%d" # YYYYMMDD
Time::DATE_FORMATS[:stamp] = "%Y%m%d%H%M%S" # YYYYMMDDHHMMSS

> ```

Reviewer: Are these built-in or ones you've created?

I'm guessing you're defining the `:stamp` symbol for both dates and times here?

Reviewer: And so, I could have changed `:stamp` to read `:foo`, right?

…then in the views.

```erb
<%= @user.last_signed_in_at.to_s(:stamp) %>
```

## Why?

> By using a method with confusing and complex arguments in your views you’re setting yourself up for inconsistency.

Reviewer: Tempted to say get rid of this sentence...but...let's leave it for now.

Add a comma after `views`.

> The main benefit for setting up global application date and time formats is for the developers who come after you. The result is a small number of consistent and pre-determined ways to present time through out your application, plus future developers can be spared building their own complicated formats!

Reviewer: Maybe you can add: 'including future you'...?

Reviewer: *throughout

Reviewer: *_and_ future developers, instead of 'plus'...

Reviewer: Maybe take the word 'complicated' out because I think you're referring to the implicit complexity that you've spoken about already? Because if the devs needed a new complicated format, then it might not be present in `time_formats.rb`, so they'd have to add it.

> Users benefit from this consistency too, if dates and times are presented in repeated ways it takes less time for them to understand. i.e. Ensuring you always say “Jun 11” rather than “11 Jun”. A small thing, but worth sweating the details.

Reviewer: ```suggestion
Users benefit from this consistency too. If dates and times are always presented in the same way, it takes less time for them to parse. I.e. Ensuring you always say “Jun 11” rather than “11 Jun”; It's a small thing, but worth sweating the details.
```

> You can also use Ruby’s `proc`s to further enhance date and time formatting where basic `strftime` usage wouldn’t be enough. See the [code for the `long_ordinal` format](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/time/conversions.rb#L15-L18) for an example.

Reviewer: Do we need this here? In this newsletter?

Why don't you do a follow-on explaining how `proc` can enhance? After all the point of this ORT is to use pre-defined formats of dates and times, not `strftime`s limitations, right?

> There’s a couple of useful sites that can help you get exactly the date formatting you prefer.

Reviewer: Colon instead of full-stop at the end.

> [For A Good Strftime](https://www.foragoodstrftime.com) contains a visually pleasant reference for `strftime` formatting strings and lets you build a live date format you can use to create new application-wide time formats.

Reviewer: ```suggestion
[For A Good Strftime](https://www.foragoodstrftime.com) contains a visually pleasant reference for `strftime` formatting strings and lets you build a live date format you can use application-wide.
```

[Strftimer](http://strftimer.com) lets you paste in a real string representation of a date of time and returns the `strftime` date format string that would produce that formatting.

> Rails also includes some defaults and there’s a site that spells out the [Rails Date and Time formats](https://railsdatetimeformats.com). Because I could never remember them and they were hard to find in the documentation.

Reviewer: I don't understand this last clause.

You're trying to say: because I couldn't remember the standard Rails Date and Time formats, I didn't know what to search for in the docs, so I had to rely on this website?

## Why not?

> There’s no real performance harm in calling `strftime` in views. It’s more about organisation and consistency.

Reviewer: ```suggestion
There’s no real performance harm in calling `strftime` in views. This is about about organisation and consistency.
```
